### PR TITLE
feat(cache): etag caching for github GET

### LINKED ISSUE #26773: Improve repo caching for unchanged data
### Describe the proposed change(s).

Take this example of a non-cloned repo:

```
DEBUG: http statistics (repository=renovate-reproductions/26724b)
       "urls": {
         "https://api.github.com/graphql (POST,200)": 4,
         "https://api.github.com/repos/renovate-reproductions/26724b/contents/renovate.json (GET,200)": 1,
         "https://api.github.com/repos/renovate-reproductions/26724b/issues/1 (GET,200)": 2,
         "https://api.github.com/repos/renovate-reproductions/26724b/pulls (GET,200)": 1
       },
       "hostStats": {"api.github.com": {"requestCount": 8, "requestAvgMs": 390, "queueAvgMs": 0}},
       "totalRequests": 8
```

Ignore the 4 graphql for now.

The next line is it fetching the contents of `renovate.json`. It knows that filename because it's cached under config.configFileName, however:
- The repo is unchanged. We should not even need to query at all if the repo's default SHA hasn't changed
- Even if the SHA has changed, we should cache the etag and reuse it so that we can get a 304 and reuse the existing contents

The next line is for the dependency dashboard issue. We should cache its content and etag to be able to get a 304.

The last line after it is for pulls. This is another case where we could cache the etag to know the list of pulls is unchanged (we already cache the content).

### PR DESCRIPTION
**Pull request by rarkins**
## Changes

Adds concept of etag caching to http layer, first used for github in these three scenarios:
- GET config file contents
- GET list of pulls
- GET issue

Example from test repo with feature enabled:

```
DEBUG: http statistics (repository=renovate-reproductions/26724b)
       "urls": {
         "https://api.github.com/graphql (POST,200)": 4,
         "https://api.github.com/repos/renovate-reproductions/26724b/contents/renovate.json (GET,304)": 1,
         "https://api.github.com/repos/renovate-reproductions/26724b/issues/1 (GET,304)": 2,
         "https://api.github.com/repos/renovate-reproductions/26724b/pulls (GET,304)": 1
       },
       "hostStats": {"api.github.com": {"requestCount": 8, "requestAvgMs": 366, "queueAvgMs": 0}},
       "totalRequests": 8
```

Note the two 304 for issues/1 is because we check the dependency dashboard again _at the end_ of a run with memCache=false in case the user has edited it during our run.

## Context

Closes #26773

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

### CODE REVIEW

**Comment by rarkins** [comment]
In GitHub we use uncacheable GraphQL for:
- init repo details
- get vulnerability alerts
- get issue list

I think the last two can be converted to REST so we can benefit from 304s

**Comment by rarkins** [comment]
@zharinov can you code review the (rare?) scenario where more than one page of PRs is needed? In such a case we don't want to cache or reuse pages 2+

File: lib/util/http/index.ts [hunk 1]
```diff
@@ -243,6 +256,34 @@ export class Http<Opts extends HttpOptions = HttpOptions> {
...
+        if (resCopy.statusCode === 200 && resCopy.headers?.etag) {
+          logger.debug(
+            `Saving response to cache: ${url} with etag ${resCopy.headers.etag}`,
+          );
+          cache.httpCache[url] = {
+            etag: resCopy.headers.etag,
+            httpResponse: copyResponse(
+              res,
+              deepCopyNeeded,
+            ) as HttpResponse<string>,
```
Comment (viceice) [inline lib/util/http/index.ts:271]:
```suggestion
            ),
```

**Comment by rarkins** [comment]
Never mind, I see that I only set repoCache=true when page idx = 1

**Review by viceice — APPROVED** [review]

**Comment by zharinov** [comment]
I don't like idea of mixing http layer with repository cache. For some requests, we may want to use package cache instead, or in-memory one even.

**Pull request merged**